src/classes/adapter.js

```diff
@@ -0,0 +1,144 @@
+   * what records are returned.
+   *
+   * @param {Object} Model - object with properties `tableName` and an `attributes` hash
+   * @return {Array} - array of database records
+   */
+  findAll (Model) {
+    const attributes = this.fieldsFromAttributes(Model.definition.attributes)
+    const tableName = Model.tableName
+    return this.knex.column(attributes).select().from(tableName)
+  }
+
+  /*
+  options should be implemented here to jsonapi spec
+  {
+    include: 'comments, posts, comments.author',
+    fields: {
+      articles: 'title, body',
+      people: 'name'
+    },
+    sort: 'age, -name',
+    page: {number: 1, size: 20},
+    filter: {}
+  }
+  */
+  query (Model, options) {
+    const attributes = this.fieldsFromAttributes(Model.definition.attributes)
+    const tableName = Model.tableName
+
+    if (typeof options.fields === 'object') {
+      const types = Object.keys(options.fields)
+      if (types.indexOf(Model.type) !== -1) {
+        const fields = options.fields[Model.type].split(',').map(field => field.trim())
+        for (const attr of attributes) {
+          if (fields.indexOf(attr) === -1) {
+            attributes.splice(attributes.indexOf(attr))
+          }
+        }
+        attributes.push('id')
+      }
+    }
+
+    let query = this.knex.column(attributes).select().from(tableName)
+
+    if (typeof options.page === 'object' && options.page.number && options.page.size) {
+      query = query
+        .limit(options.page.size)
+        .offset(options.page.number * options.page.size - options.page.size)
+    }
+
+    if (typeof options.sort === 'string') {
+      options.sort.split(',').forEach(column => {
+        const direction = (column.indexOf('-') !== -1) ? 'DESC' : 'ASC'
+        column = column.replace('-', '')
+        query = query.orderBy(column.trim(), direction)
+      })
+    }
+
+    if (typeof options.include === 'string') {
+      const includes = this.parseIncludes(options.include)
+      for (const include of includes) {
+        // needs to recurse since we may need to get comments for posts but also
+        // author for comments
+        // I think the include is going to need to perform queries, create
+        // lists and then merge them
+        this.include(query, include)
+      }
+    }
+
+    return query
+  }
+
+  /**
+   * Retrieves a single record for the given model by given id
+   *
+   * @param {Object} Model - object with properties `tableName` and an `attributes` hash
+   * @param {number} id - id of record to retrieve
+   * @return {Object} - a single record or null if no record was found
+   */
+  findRecord (Model, id) {
+    const attributes = this.fieldsFromAttributes(Model.definition.attributes)
+    const tableName = Model.tableName
+    return this.knex.column(attributes).first().from(tableName).where({id}).limit(1)
+      .then(result => result || null)
+  }
+
+  queryRecord (Model, options) {
+    return this.query(Model, options).first().limit(1)
```

Comment:
Why `first().limit(1)`? You should only need `.first()` right?

6/10 careless code